File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        return maxDepth(root)[2];
    }

    public int[] maxDepth(TreeNode root){
        if (null==root){
            return new int[]{0,0,0};
        }

        int hl = 0, hr = 0, dl = 0, dr = 0;
        if (null!=root.left){
            int[] res = maxDepth(root.left);
            hl = Math.max(res[0], res[1])+1;
            dl = res[2];
        }
        if (null!=root.right){
            int[] res = maxDepth(root.right);
            hr = Math.max(res[0], res[1])+1;
            dr = res[2];
        }

        return new int[]{hl, hr, Math.max(Math.max(dl, dr),hl+hr)};
        
    }
}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        return self.maxHeight(root)[2]

    def maxHeight(self, root: Optional[TreeNode])->List[int]:
        if not root:
            return [0,0,0]
        hl, hr, ld, lr= 0,0,0,0 
        if root.left:
            lhl, lhr, ld = self.maxHeight(root.left)
            hl = max(lhl,lhr) + 1
        
        if root.right:
            rhl, rhr, lr = self.maxHeight(root.right)
            hr = max(rhl, rhr) + 1
        
        # third arg is the diameter
        # it will be max of diameter of left, right subtree and subtree from current root
        return [hl, hr, max(hl+hr, ld, lr)]